### ISSUE TITLE
<<- vs <- assignment

### ISSUE THREAD

**Issue by christophergandrud** [IQSS/Zelig#172]
This regards the series of build warnings related where R urges Zelig to use deep assignment.

CAVEAT: I talked with Christine about this a few weeks ago, got caught up with other things and am just returning to it now, so I'm probably forgetting something.

What is the rational for not doing this? I put together a [simple test](https://github.com/IQSS/Zelig/tree/refClassAssignment) where `num` in `sim` is assigned using deep assignment. This seems to produce expected behaviour, e.g.:

```R
devtools::install_github('IQSS/Zelig', 'refClassAssignment')

library(Zelig)
z5 <- zls$new()
z5$zelig(Fertility ~ Education, data = swiss)
z5$setx(Education = 5)

z5$sim()
length(z5$getqi())
## [1] 1000

z5$sim(num = 10)
length(z5$getqi())
## [1] 10

z5$sim()
length(z5$getqi())
## [1] 10
``` 

What am I missing here?

**Comment by cchoirat** [comment]
Sorry, I'm not sure I get your concern:

num <<- value

and

.self$num <- value

behave the same (if num is a field in the object).

**Comment by christophergandrud** [comment]
I'm just trying to remember why:

- https://github.com/IQSS/Zelig/blob/master/R/model-zelig.R#L530

Instead of:

- https://github.com/IQSS/Zelig/blob/refClassAssignment/R/model-zelig.R#L530

If the latter means that a warning isn't produced at build.

**Comment by cchoirat** [comment]
There's a warning in this case because assignment is ambiguous (is num the
object field or a local variable?).  See, e.g., localauthors vs authors.

**Comment by christophergandrud** [comment]
So a preferred solution to using deep assignment would be:

```R
local_num <- 1000
.self$num <- local_num
```

rather than:

```R
num <<- 1000
```

Is that right?

**Comment by cchoirat** [comment]
No, no, I'm just saying that:

(1)

local_num <- value1
.self$num <- value2

and

(2)

local_num <- value1num <<- value2

are the same

and that warnings come from statements such as:

(3)

num <- value

**Comment by christophergandrud** [comment]
Right. Is there any reason why I should not replace all:

```R
    if (length(.self$num) == 0){
      if(is.null(num)){
        num <- 1000
      }
    }
    if(!is.null(num)){
      .self$num <- num
    }
```

with

```R
    if (length(.self$num) == 0){
      if(is.null(num)){
        local_num <- 1000
        .self$num <- local_num
      }
    }
    if(!is.null(num)){
      .self$num <- num
    }
```

or alternatively (as suggested by R):

```R
    if (length(.self$num) == 0){
      if(is.null(num)){
        num <<- 1000
      }
    }
    if(!is.null(num)){
      .self$num <- num
    }
```

All seem to have the same behaviour, the latter two produce no warnings.

**Comment by christophergandrud** [comment]
I've implemented the deep assignment operator for `num` in 5.0-14. 

I'll work on the other ones for 5.0-15.

**Comment by tercer** [comment]
Given that addressing this has created a few tangential issues, I used adjustment of local names to avoid the CRAN warnings that were admittedly quite annoying.  

Here is install from source for me now:

<img width="619" alt="screen shot 2017-02-06 at 1 34 11 pm" src="https://cloud.githubusercontent.com/assets/2085037/22661606/f0e40e56-ec73-11e6-96a6-434d39d82e2f.png">

**Comment by christophergandrud** [comment]
Also resolved in ZeligChoice for [version 0.9-4](https://github.com/IQSS/ZeligChoice/commit/000322df56cb4c7b574687a2a24faf9ba615943b)
